pkg/importer/dataStream.go

```diff
@@ -182,31 +222,71 @@ func (d *dataStream) constructReaders() ([]io.ReadCloser, error) {
 	}
 
 	// if image is qemu convert it to raw
-	magicStr, err := image.GetMagicNumber(rdr)
+	magicStr, err := image.GetMagicNumber(rdr.Rdr)
 	if err != nil {
 		return nil, errors.WithMessage(err, "unable to check magic number")
 	}
 	d.Qemu = image.MatchQcow2MagicNum(magicStr)
 	// don't lose bytes reading the magic number regardless of the file being qemu or not:
 	// MultiReader reads from each reader, in order, until the last reader returns eof.
-	multir := io.MultiReader(bytes.NewReader(magicStr), rdr)
-	readers = append(readers, ioutil.NopCloser(multir))
+	multir := io.MultiReader(bytes.NewReader(magicStr), rdr.Rdr)
+	readers = append(readers, Reader{RdrType: RdrMulti, Rdr: ioutil.NopCloser(multir)})
 
 	return readers, nil
 }
 
-// Close the passed-in slice of readers in reverse order. See constructReaders().
-func CloseReaders(readers []io.ReadCloser) {
-	for i := len(readers) - 1; i >= 0; i-- {
-		r := readers[i]
-		r.Close()
+func GzReader(r io.ReadCloser) (Reader, error) {
+	gz, err := gzip.NewReader(r)
+	if err != nil {
+		return Reader{}, errors.Wrap(err, "could not create gzip reader")
+	}
+	glog.V(Vadmin).Infof("gzip: extracting %q\n", gz.Name)
+	return Reader{RdrType: RdrGz, Rdr: gz}, nil
+}
+
+func XzReader(r io.ReadCloser) (Reader, error) {
+	glog.V(Vdebug).Infoln("XzReader: xz format")
+	xz, err := xz.NewReader(r, 0) //note: default dict size may be too small
+	if err != nil {
+		return Reader{}, errors.Wrap(err, "could not create xz reader")
+	}
+	return Reader{RdrType: RdrXz, Rdr: ioutil.NopCloser(xz)}, nil
+}
+
+func TarReader(r io.ReadCloser) (Reader, error) {
+	tr := tar.NewReader(r)
+	hdr, err := tr.Next() // advance cursor to 1st (and only) file in tarball
+	if err != nil {
+		return Reader{}, errors.Wrap(err, "could not read tar header")
+	}
+	glog.V(Vadmin).Infof("tar: extracting %q\n", hdr.Name)
+	return Reader{RdrType: RdrTar, Rdr: ioutil.NopCloser(tr)}, nil
+}
+
+// Close the passed-in slice in order except the 1st reader is closed last, see constructReaders().
+func closeReaders(readers []Reader) (rtnerr error) {
+	if len(readers) <= 1 {
+		// no readers to close (multi-reader does not need to be closed)
+		return nil
+	}
+	var err error
+	// move first reader to end of slice
+	r := readers[0]
+	readers = readers[1:]
+	readers = append(readers, r)
+	for _, r = range readers {
```

Comment:
This closes a .tar.gz reader stack as (gz ->  tar -> http).  This is not the order the pipeline is structured (http -> gz -> tar).  Is the intent to close the readers in reverse order?

Reply:
+1